src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs

```diff
@@ -555,26 +667,46 @@ private void RunDebugExecutionLoop()
             try
             {
                 DebugContext.EnterDebugLoop();
-                RunExecutionLoop();
+                RunExecutionLoop(isForDebug: true);
             }
             finally
             {
                 DebugContext.ExitDebugLoop();
             }
         }
 
-        private void RunExecutionLoop()
+        private void RunExecutionLoop(bool isForDebug = false)
         {
+            Runspace initialRunspace = Runspace;
             while (!ShouldExitExecutionLoop)
             {
-                using CancellationScope cancellationScope = _cancellationContext.EnterScope(isIdleScope: false);
+                if (isForDebug && !initialRunspace.RunspaceStateInfo.IsUsable())
+                {
+                    return;
+                }
+
+                using CancellationScope cancellationScope = _cancellationContext.EnterScope(false);
                 DoOneRepl(cancellationScope.CancellationToken);
 
                 while (!ShouldExitExecutionLoop
                     && !cancellationScope.CancellationToken.IsCancellationRequested
                     && _taskQueue.TryTake(out ISynchronousTask task))
                 {
                     task.ExecuteSynchronously(cancellationScope.CancellationToken);
+                    while (Runspace is { RunspaceIsRemote: true } remoteRunspace
+                        && !remoteRunspace.RunspaceStateInfo.IsUsable())
+                    {
+                        PopPowerShell(RunspaceChangeAction.Exit);
+                    }
+                }
+
+                if (_shouldExit
+                    && (CurrentFrame.FrameType & PowerShellFrameType.Remote) is not 0
+                    && (CurrentFrame.FrameType & PowerShellFrameType.Repl) is not 0
+                    && (CurrentFrame.FrameType & PowerShellFrameType.Nested) is 0)
+                {
+                    _shouldExit = false;
+                    PopPowerShell();
```

Comment:
Ok, we'll need to walk through the logic here and above together, it hurt my head 🤕

Reply:
It hurt my head to write it if it makes you feel any better 😁 

For real though, a lot of what made PSES complicated before was controlling for scenarios like this. One runspace is pretty easy, multiple is a lot harder. It's also what made it prone to dead locks and slowness.

The difference this time though is I wrote it in a way where the extra complexity only really has an effect when it has to. So the normal everyday debugging loop should be completely unaffected but we still get reliable attach debugging. Thankfully Rob's work made it *much* easier to write in this way.

Reply:
🥳